pkg/services/notary/notary.go

```diff
@@ -0,0 +1,368 @@
+				n.Config.Log.Error("failed to finalize main transaction", zap.Error(err))
+			} else {
+				r.isSent = true
+			}
+			continue
+		}
+		if r.minNotValidBefore <= bc.BlockHeight() { // then at least one of the fallbacks can already be sent.
+			newFallbacks := r.fallbacks[:0]
+			for _, fb := range r.fallbacks {
+				if nvb := fb.GetAttributes(transaction.NotValidBeforeT)[0].Value.(*transaction.NotValidBefore).Height; nvb <= bc.BlockHeight() {
+					if err := n.finalize(fb); err != nil {
+						newFallbacks = append(newFallbacks, fb) // wait for the next block to resend them
+					}
+				} else {
+					newFallbacks = append(newFallbacks, fb)
+				}
+			}
+			if len(newFallbacks) == 0 {
+				delete(n.requests, h)
+			} else {
+				r.fallbacks = newFallbacks
+			}
+		}
+	}
+}
+
+// finalize adds missing Notary witnesses to the transaction (main or fallback) and pushes it to the network.
+func (n *Notary) finalize(tx *transaction.Transaction) error {
+	acc := n.getAccount()
+	if acc == nil {
+		panic(errors.New("no available Notary account"))
+	}
+	notaryWitness := transaction.Witness{
+		InvocationScript:   append([]byte{byte(opcode.PUSHDATA1), 64}, acc.PrivateKey().Sign(tx.GetSignedPart())...),
+		VerificationScript: []byte{},
+	}
+	for i, signer := range tx.Signers {
+		if signer.Account == n.Config.Chain.GetNotaryContractScriptHash() {
+			tx.Scripts[i] = notaryWitness
+			break
+		}
+	}
+	return n.getOnTransaction()(tx)
+}
+
+// SetOnTransaction sets callback method which should be called after transaction is completed.
+func (n *Notary) SetOnTransaction(f func(tx *transaction.Transaction) error) {
+	n.mtx.Lock()
+	defer n.mtx.Unlock()
+	n.onTransaction = f
+}
+
+func (n *Notary) getOnTransaction() func(tx *transaction.Transaction) error {
+	n.mtx.RLock()
+	defer n.mtx.RUnlock()
+	return n.onTransaction
+}
+
+// verifyIncompleteWitnesses checks that tx either doesn't have all witnesses attached (in this case none of them
+// can be multisignature), or it only has a partial multisignature. It returns the request type (sig/multisig), the
+// number of signatures to be collected, sorted public keys (for multisig request only) and an error.
+func (n *Notary) verifyIncompleteWitnesses(t *transaction.Transaction, nKeys uint8) (RequestType, uint8, keys.PublicKeys, error) {
+	var (
+		typ         RequestType
+		nSigs       int
+		nKeysActual uint8
+		pubsBytes   [][]byte
+		pubs        keys.PublicKeys
+		ok          bool
+	)
+	for i, w := range t.Scripts {
+		// do not check witness for Notary contract -- it will be replaced by proper witness in any case.
+		if t.Signers[i].Account == n.Config.Chain.GetNotaryContractScriptHash() {
+			continue
+		}
+		if len(w.VerificationScript) == 0 { // then it's a dummy signature witness, don't need to check invocation script
+			if typ == MultiSignature {
+				return Unknown, 0, nil, fmt.Errorf("bad type of witness #%d: multisignature witness can not be combined with other witnesses", i)
```

Comment:
...which can be successful even without any invocation script (think of oracle response) and can be safely combined with anything.

Reply:
Fixed, now empty witnesses are alowed.